include/ginkgo/core/base/mpi.hpp

```diff
@@ -0,0 +1,471 @@
+namespace gko {
+namespace mpi {
+
+enum class op_type {
+    sum = 1,
+    min = 2,
+    max = 3,
+    product = 4,
+    custom = 5,
+    logical_and = 6,
+    bitwise_and = 7,
+    logical_or = 8,
+    bitwise_or = 9,
+    logical_xor = 10,
+    bitwise_xor = 11,
+    max_val_and_loc = 12,
+    min_val_and_loc = 13
+};
+
+
+/*
+ * Class that allows an RAII of initialization and calls MPI_Finalize at the
+ * end of its scope. Therefore this must be called before any of the MPI
+ * functions.
+ */
+class init_finalize {
+public:
+    init_finalize(int& argc, char**& argv, const size_type num_threads = 1);
+
+    init_finalize() = delete;
+
+    init_finalize(init_finalize& other) = default;
+
+    init_finalize& operator=(const init_finalize& other) = default;
+
+    init_finalize(init_finalize&& other) = default;
+
+    init_finalize& operator=(init_finalize&& other) = default;
+
+    static bool is_finalized();
+
+    static bool is_initialized();
+
+    ~init_finalize();
+
+private:
+    int num_args_;
+    int required_thread_support_;
+    int provided_thread_support_;
+    char** args_;
+};
+
+
+/**
+ * A class holding and operating on the MPI_Info class. Stores the key value
+ * pair as a map and provides methods to access these values with keys as
+ * strings.
+ */
+class info {
+public:
+    info();
+
+    info(MPI_Info input) { this->info_ = input; }
+
+    void remove(std::string key);
+
+    std::string& at(std::string& key) { return this->key_value_.at(key); }
+
+    void add(std::string key, std::string value);
+
+    MPI_Info get() { return this->info_; }
+
+    ~info();
+
+private:
+    std::map<std::string, std::string> key_value_;
+    MPI_Info info_;
+};
+
+
+/**
+ * A request class that takes in the given request and duplicates it
+ * for our purposes. As the class or object goes out of scope, the request
+ * is freed.
+ */
+class request : public EnableSharedCreateMethod<request> {
+public:
+    request(const int size) : req_(new MPI_Request[size]) {}
+
+    request() : req_(new MPI_Request[1]) {}
+
+    ~request()
+    {
+        if (req_) delete[] req_;
+    }
+
+    MPI_Request* get_requests() const { return req_; }
+
+private:
+    MPI_Request* req_;
+};
+
+
+/**
+ * A status class that takes in the given status and duplicates it
+ * for our purposes. As the class or object goes out of scope, the status
+ * is freed.
+ */
+class status : public EnableSharedCreateMethod<status> {
+public:
+    status(const int size) : status_(new MPI_Status[size]) {}
+
+    status() : status_(new MPI_Status[1]) {}
+
+    ~status()
+    {
+        if (status_) delete[] status_;
+    }
+
+    MPI_Status* get_statuses() const { return status_; }
+
+private:
+    MPI_Status* status_;
+};
+
+
+/**
+ * A communicator class that takes in the given communicator and duplicates it
+ * for our purposes. As the class or object goes out of scope, the communicator
+ * is freed.
+ */
+class communicator : public EnableSharedCreateMethod<communicator> {
```

Comment:
I guess, I don't understand how you imagine the type with value semantics then. Could you be a bit more direct? I'm especially wondering, how the `MPI_Comm` should be encapsulated. Storing it by value has the disadvantage that I've mentioned above, so it would need to be something pointer/reference like. Using a pointer seems weird, as I don't think you would generally use it that way except to create comms.

Reply:
```cpp
struct communicator {
    int rank;
    int size;
    MPI_Comm comm;
}
```
without any special semantics attached. The only reason we can't use MPI_Comm directly is that we cache size and rank.

Reply:
As Terry pointed out, MPI_Comm_rank and MPI_Comm_size are really lightweight functions in both OpenMPI and MPICH, I don't think it makes sense to cache their values here.